### PR TITLE
[Merged by Bors] - fix: issues in linear_combination + polyrith

### PR DESCRIPTION
**Pull request by digama0**
As [reported on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Bug.20in.20.60polyrith.60.3F/near/339764147).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you summarize why this fixes things?

**Comment by digama0** [comment]
There are three separate bugs being fixed here.

1. The argument order change in `c_add_pf` helps the elaborator to determine `a : A` based on the type of `h`, especially when coercions are involved. This was the original cause of the bug report.
2. The change to use `case` in `elabLinearCombination` is fixing a bug wherein the typeclass goals thrown up by (1) are also solved by `ring`, causing even more strange errors.
3. The change to use `withRef` in `polyrith` is fixing a bug wherein the errors reported in (1) and (2) appear at location 0:0 (i.e. the top of the file) which is very confusing. Now the errors are reported on the span of `"polyrith"` as you would expect.

**Comment by semorrison** [comment]
bors merge
